### PR TITLE
feat(Core/Battlegrounds) Reworked enhanced bg system for modules

### PR DESCRIPTION
**Pull request by Shard-MW**
Reworked https://github.com/azerothcore/azerothcore-wotlk/pull/2334
Special thanks to for his main PR @Helias <3

##### CHANGES PROPOSED:
- Fixed score errors from base PR (while queuing Random Queue)
- Fixed worldstate/score errors in our bg-modules that were occurring only from Random Queue
- Removed non-needed changes
- Added hashmaps and conditions

##### TESTS PERFORMED:
- Joined and played until end every battlegrounds from Random Queue
- Joined and played until end every battlegrounds from Specific Queue
- Joined and played until end every arenas (skirmish/rated)
- Checked score & worldstates everytime

##### HOW TO TEST THE CHANGES:
- Join & play every battlegrounds from Random Queue
- Join & play every battlegrounds from Specific Queue
- Join & play every arenas (skirmish/rated)
- Don't forget to check scores panel

##### Target branch(es):
- [x] Master

## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

**Comment by FrancescoBorzi** [comment]
@Endeffect-AT can you please have a look at this?

**Comment by Endeffect-AT** [comment]
Will test this and PR # 1903 the next days, just need a little time

**Comment by Endeffect-AT** [comment]
I guess it's wip and i test it when ready 👍

**Comment by Shard-MW** [comment]
> I guess it's wip and i test it when ready 👍
@Endeffect-AT 
It's not WIP, it's a finished PR haha

File: src/server/game/Battlegrounds/BattlegroundMgr.cpp [hunk 1]
```diff
@@ -1091,3 +1099,19 @@ std::unordered_map<int, bgRef> BattlegroundMgr::bgTypeToTemplate = {
     { BATTLEGROUND_RB, [](Battleground *bg_t) -> Battleground*{ return new Battleground(*bg_t); }, },
     { BATTLEGROUND_AA, [](Battleground *bg_t) -> Battleground*{ return new Battleground(*bg_t); }, },
 };
+
+std::unordered_map<int, bgMapRef> BattlegroundMgr::getBgFromMap = {};
+
+std::unordered_map<int, bgTypeRef> BattlegroundMgr::getBgFromTypeID = {
```
[the remark this answers is not in the document]
Comment (Shard-MW) [reply src/server/game/Battlegrounds/BattlegroundMgr.cpp]: Why ?

This is the current AzerothCore code : 
https://github.com/azerothcore/azerothcore-wotlk/blob/master/src/server/game/Battlegrounds/BattlegroundMgr.cpp#L1031

https://github.com/azerothcore/azerothcore-wotlk/blob/master/src/server/game/Battlegrounds/BattlegroundMgr.cpp#L1048

(...)

**Comment by Endeffect-AT** [comment]
I refer to the requested changes from @Viste

File: src/server/game/Battlegrounds/BattlegroundMgr.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Viste) [reply src/server/game/Battlegrounds/BattlegroundMgr.cpp]: I don't know when it was added, but what you're referring to is also wrong.

**Comment by Endeffect-AT** [comment]
Tested on win 10 pro 64 bit, AC commit 70170f0. 
Ran build without any errors, no crashes during the performed tests. Every bg was tested with .debug bg and one pleayer each faction. 

Note: Didn't check the worldstates.


Battlegrounds
---------------------------------------------

**Joined & played every BG until the end from Random Queue:**

- Isle of Conquest 
> Scoreboard Error: Joined with one pleayer each faction, 2 Horde are shown. 

- Eye of the Storm
> Statistics work as expected	

- Alterac Valley
> Statistics work as expected

- Strand of the Ancients
> Statistics work as expected

- Warsong Gulch
> Statistics work as expected

- Arathi Basin
> Statistics work as expected
> - Farm Horde Banner Bug: As horde char you didn't see and can't interactive with the banner. Only as the Alliance has taken it, you could'd see and interact with it. 

**Joined & played every BG until the end from Specific Queue:**

- Isle of Conquest 
> Statistics work as expected

- Eye of the Storm
> Statistics work as expected

- Alterac Valley
> Statistics work as expected

- Strand of the Ancients
> Statistics work as expected

- Warsong Gulch
> Statistics work as expected

- Arathi Basin
> Statistics work as expected


Arenas rated (2v2)
---------------------------------------------

Team with ingame petition signed and joined every arena:

- Nagrand Arena
> Statistics work as expected

- Ring of Valor
> Statistics work as expected

- Dalaran Arena
> Statistics work as expected

- Ruins of Lordaeron
> Statistics work as expected

- Circle of Blood Arena
> Statistics work as expected


Arenas skirmish (2v2)
---------------------------------------------

Joined every arena in group:

- Ring of Valor
> Statistics work as expected

- Circle of Blood Arena
> Statistics work as expected

- Ruins of Lordaeron
> Statistics work as expected

- Nagrand Arena
> Statistics work as expected

- Dalaran Arena
> Statistics work as expected


**ToDo - List:**
---------------------------------------------

- [ ] Fix Scoreboard in BG "Isle of Conquest" from Random Queue
- [ ] Fix Horde Farm Banner in "Arathi Basin" from Random Queue


Others:
---------------------------------------------

- Wintergrasp (Didn't played until end, just joined to test)
> If you join Wintergrasp on both sides in Dalaran at the Battlemaster an click on minimap "Leave Wintergrasp" you got teleported at the starting zone of your race. I don't think this should happen....

**Comment by Shard-MW** [comment]
Thanks a lot for taking that much time for testing, you're awesome @Endeffect-AT ! <3

It's weird that only "Isle of Conquest" have an issue with scoreboard, since this PR shouldn't touch at this... Can you try if you've the same error without the PR please ?

About
> Fix Horde Farm Banner in "Arathi Basin" from Random Queue
this is not triggered by the PR. It's a random bug that AC have, I've reported this before :
https://github.com/azerothcore/azerothcore-wotlk/issues/2518

**Comment by FrancescoBorzi** [comment]
@Endeffect-AT great job! What a nice report. 

I don't think that any of the bugs that you mentioned have been introduced by this PR. 

For example, the Arathi Basin missing flag is an ancient bug that randomly happens and I usually clean my client cache to fix it

**Comment by Endeffect-AT** [comment]
Thx guys <3

I noticed the bug with the banner the first time on testing  this:
https://github.com/azerothcore/azerothcore-wotlk/pull/1903
I thought it was caused by me witg GM Mode. I tried it again and it worked. Okay so this is not related to this PR!

I will test the Scoreboard in master soon @!

**Comment by Endeffect-AT** [comment]
Tested Isle of Conquest from Random Queue on master commit 0f590d03974a+ (Yes, i should update my core...) and the bug is the same. 

image

So your PR is working fine imo 👍

**Comment by Shard-MW** [comment]
@Endeffect-AT 
Thanks again ! I think you can open an issue about Isle of Conquest so :P

**Comment by FrancescoBorzi** [comment]
@Endeffect-AT cool, thanks for checking that too. And good job @Shard-MW

**Comment by Viste** [comment]
plz fix codestyle.

**Comment by Shard-MW** [comment]
> plz fix codestyle.

Done, but I'll not change this : 
https://github.com/azerothcore/azerothcore-wotlk/pull/2521/files#diff-85ba4e80476ac43de064d54592507596R1105
because it looks horrible since it's not a function

**Pull request merged**